z-library-desktop: init at 3.1.0

UlyssesZh opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

UlyssesZh committed:
z-library-desktop: init at 3.1.0

awwpotato commented on pkgs/by-name/zl/zlibrary-desktop/package.nix:
```diff
@@ -0,0 +1,201 @@
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = pname;
+      desktopName = "Z-Library";
+      comment = meta.description;
```
these are both static strings that are unlikely to change, see https://github.com/NixOS/nixpkgs/issues/277994
```suggestion
      name = "zlibrary-desktop";
      desktopName = "Z-Library";
      comment = "client for the online library Z-Library;
```

awwpotato commented on pkgs/by-name/zl/zlibrary-desktop/package.nix:
```diff
@@ -0,0 +1,201 @@
+    ]);
+in
+stdenv.mkDerivation {
+  inherit pname version meta;
```
all of these can be inlined when my other comment is resolved

awwpotato commented on pkgs/by-name/zl/zlibrary-desktop/package.nix:
```diff
@@ -0,0 +1,198 @@
+}:
+
+let
+  stdenv = stdenvNoCC;
```
nit: is there any reason not to inline this?

UlyssesZh replied:
Because I am lazy to write `NoCC` every time? I am thinking of possible future packaging on darwin, which would need `stdenv.hostPlatform` etc., making it worth abbreviating this `NoCC`. Also in case we need cc in the future, it would be easier to change that.

awwpotato commented on pkgs/by-name/zl/zlibrary-desktop/package.nix:
```diff
@@ -0,0 +1,198 @@
+
+let
+  stdenv = stdenvNoCC;
+  libraries =
```
I would inline this, but it doesn't really matter, up to you.

UlyssesZh replied:
The variable `libraries` is used twice. I would have to write the long list twice if it is inline.

awwpotato replied:
oh oops didn't notice that

awwpotato commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 399610`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>zlibrary-desktop</li>
  </ul>
</details>

awwpotato commented on pkgs/by-name/zl/zlibrary-desktop/package.nix:
```diff
@@ -0,0 +1,198 @@
+
+      runHook postInstall
+    '';
+
```
```suggestion

  passthru.updateScript = nix-update-script { };

```

UlyssesZh replied:
This package would need a handwritten update script if there is one, and it will involve using API of Wayback Machine to get a snapshot of the src. Because the address of Z-Library's website is always changing (except the onion one), writing an update script would be hard. Therefore, I decide not to include an update script. This package, if merged, will be updated on request or when the package maintainer notice that there is an update.

awwpotato replied:
Sorry, I forgot about the fact this isn't from a git repo :facepalm:. Totally reasonable not to include an update script in this case.

nadir-ishiguro commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 399610`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>zlibrary-desktop</li>
  </ul>
</details>

nadir-ishiguro commented:
Didn't even know about this. Thank your packaging this. 

During testing, I couldn't download books:

`(z-library:34490): GLib-GIO-ERROR **: 14:32:44.522: Settings schema 'org.gtk.Settings.FileChooser' is not installed`

The FileChooser is definitely available, so I assume this is somehow due to the testing environment, but I thought I'd mention it, as I don't remember having that issue with other graphical applications in nixpkgs-review.

nadir-ishiguro commented:
Okay, narrowed it down a bit. 

I forgot to mention that the original test was in a Niri session.

Just installed it in an Archlinux distrobox via the AUR and yep, FileChooser works as expected.

In GNOME it works with both this derivation, started from the nixpkgs-review environment, and with the distrobox/Arch package.

UlyssesZh commented:
Updated: 2.4.1 -> 2.4.2.

Added an update script. The scripts find the latest URL of Z-Library (which is always changing) by reading from a Wikipedia article. It also makes an HTTP request to Wayback Machine to get an archive link.

UlyssesZh commented:
Use Electron from nixpkgs to significantly reduce package size and also make it available for more platforms. However it cannot build on darwin because `dpkg` is marked broken (https://github.com/NixOS/nixpkgs/pull/370245#discussion_r2126471012).

UlyssesZh commented on pkgs/by-name/z-/z-library-desktop/package.nix:
```diff
@@ -0,0 +1,59 @@
+    makeWrapper
+    asar
+  ];
+
```
It's not necessary because string interpolation will force it to build.

UlyssesZh commented on pkgs/by-name/z-/z-library-desktop/package.nix:
```diff
@@ -0,0 +1,59 @@
+    mkdir -p $phome
+    cp -r opt/Z-Library/resources $phome
+    pushd $phome/resources
+    asar e app.asar app
```
There are symlinks.

marcin-serwin reviewed (approved):
Builds and runs on `x86_64-linux`.

UlyssesZh commented on pkgs/by-name/z-/z-library-desktop/package.nix:
```diff
@@ -0,0 +1,90 @@
+    homepage = "https://z-library.sk";
+    description = "client for the online library Z-Library";
+    license = lib.licenses.unfree; # Maintainers on AUR emailed the dev to confirm: https://pastebin.com/ss4Nr8pW
+    platforms = lib.platforms.all;
```
Actually I can probably make it work by building the keytar library ourselves.

UlyssesZh commented:
Tested that it is usable on darwin too now.

SuperSandro2000 commented on pkgs/by-name/z-/z-library-desktop/package.nix:
```diff
@@ -0,0 +1,138 @@
+    src = fetchFromGitHub {
+      owner = "atom";
+      repo = "node-keytar";
+      rev = "v${finalAttrs.version}";
```
```suggestion
      tag = "v${finalAttrs.version}";
```

SuperSandro2000 commented on pkgs/by-name/z-/z-library-desktop/update.rb:
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i ruby -p ruby ruby.gems.nokogiri
```
Can we use shell to be similar to the standard update scripts?

UlyssesZh commented on pkgs/by-name/z-/z-library-desktop/update.rb:
```diff
@@ -0,0 +1,79 @@
+finish 'Version is up to date.' if version == current_version
+abort "Could not update version: #{current_version} -> #{version}" unless nix.sub! current_version, version
+
+download = item.at_css('a.za-download-href-link[href$=".deb"]')&.[] 'href'
```
If you consider that the page may change, using regex is just as unreliable (if not more) as CSS selectors. They are also less readable and maintainable than CSS selectors in my opinion.
